File: tfjs-backend-wasm/src/cc/kernels/FusedDepthwiseConv2D_test.cc
/* Copyright 2019 Google Inc. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ===========================================================================*/

#include <gtest/gtest.h>
#include <xnnpack.h>

#include "src/cc/backend.h"
#include "src/cc/conv2d_impl.h"
#include "src/cc/kernels/FusedDepthwiseConv2D.h"
#include "src/cc/util.h"

TEST(FUSEDDEPTHWISECONV2D, xnn_operator_lifetime) {
  tfjs::wasm::init();

  ASSERT_EQ(0, tfjs::backend::num_tensors());

  const int x0_id = 0;
  const int x1_id = 1;
  const int size = 8;
  float x_values[size] = {1, 2, 3, 4, 5, 6, 7, 8};

  const int weights0_id = 2;
  const int weights1_id = 3;
  const int weights_size = 8;
  float weights_values[weights_size] = {1, 2, 3, 4, 5, 6, 7, 8};

  const int bias0_id = 4;
  const int bias1_id = 5;
  const int bias_size = 1;
  float bias_values[bias_size] = {1};

  const int out_id = 6;
  const int out_size = 12;
  float out_values[out_size] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};

  tfjs::wasm::register_tensor(x0_id, size, x_values);
  tfjs::wasm::register_tensor(x1_id, size, x_values);
  tfjs::wasm::register_tensor(weights0_id, weights_size, weights_values);
  tfjs::wasm::register_tensor(weights1_id, weights_size, weights_values);
  tfjs::wasm::register_tensor(bias0_id, bias_size, bias_values);
  tfjs::wasm::register_tensor(bias1_id, bias_size, bias_values);
  tfjs::wasm::register_tensor(out_id, out_size, out_values);

  ASSERT_EQ(7, tfjs::backend::num_tensors());
  ASSERT_EQ(0, tfjs::backend::xnn_operator_count);

  // One xnn_operator should be created for the first call to conv2d with no
  // bias.
  const int batch_size = 1;
  const int input_height = 4;
  const int input_width = 2;
  const int filter_height = 4;
  const int filter_width = 2;
  const int pad_top0 = 1;
  const int pad_right = 0;
  const int pad_bottom0 = 0;
  const int pad_left = 0;
  const bool is_same_pad0 = false;
  const int dilation_height = 1;
  const int dilation_width = 1;
  const int stride_height = 1;
  const int stride_width = 1;
  const int input_channels = 1;
  const int output_channels = 1;

  const int activation = tfjs::wasm::FusableActivation::LINEAR;

  tfjs::wasm::FusedDepthwiseConv2D(
      x0_id, batch_size, input_height, input_width, weights0_id, filter_height,
      filter_width, -1 /* bias */, pad_top0, pad_right, pad_bottom0, pad_left,
      is_same_pad0, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, activation,
      -1 /* prelu weights */, out_id);
  ASSERT_EQ(1, tfjs::backend::xnn_operator_count);

  // One new xnn operator should be created for second call to conv2d with no
  // bias and prelu activation.
  const int prelu_activation = tfjs::wasm::FusableActivation::PRELU;

  const int prelu_weights_id = 7;
  const int prelu_size = 8;
  float prelu_values[prelu_size] = {1, 2, 3, 4, 5, 6, 7, 8};
  tfjs::wasm::register_tensor(prelu_weights_id, prelu_size, prelu_values);

  tfjs::wasm::FusedDepthwiseConv2D(
      x0_id, batch_size, input_height, input_width, weights0_id, filter_height,
      filter_width, -1 /* bias */, pad_top0, pad_right, pad_bottom0, pad_left,
      is_same_pad0, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, prelu_activation,
      prelu_weights_id, out_id);
  ASSERT_EQ(2, tfjs::backend::xnn_operator_count);

  // No new xnn_operators should be created for the second call to conv2d with
  // the same arguments.
  tfjs::wasm::FusedDepthwiseConv2D(
      x0_id, batch_size, input_height, input_width, weights0_id, filter_height,
      filter_width, -1 /* bias */, pad_top0, pad_right, pad_bottom0, pad_left,
      is_same_pad0, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, activation,
      prelu_weights_id, out_id);
  ASSERT_EQ(2, tfjs::backend::xnn_operator_count);

  // No new xnn_operators should be created for the second call to conv2d with
  // the same arguments but different input.
  tfjs::wasm::FusedDepthwiseConv2D(
      x1_id, batch_size, input_height, input_width, weights0_id, filter_height,
      filter_width, -1 /* bias */, pad_top0, pad_right, pad_bottom0, pad_left,
      is_same_pad0, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, activation,
      prelu_weights_id, out_id);
  ASSERT_EQ(2, tfjs::backend::xnn_operator_count);

  // One new xnn_operator should be created for the next call to conv2d with the
  // same weights and bias but different arguments.
  const int pad_top1 = 0;
  const int pad_bottom1 = 1;
  tfjs::wasm::FusedDepthwiseConv2D(
      x0_id, batch_size, input_height, input_width, weights0_id, filter_height,
      filter_width, -1 /* bias */, pad_top1, pad_right, pad_bottom1, pad_left,
      is_same_pad0, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, activation,
      prelu_weights_id, out_id);
  ASSERT_EQ(3, tfjs::backend::xnn_operator_count);

  // One more xnn operator should be created for the next call to conv2d with
  // new weights and same input.
  tfjs::wasm::FusedDepthwiseConv2D(
      x0_id, batch_size, input_height, input_width, weights1_id, filter_height,
      filter_width, -1 /* bias */, pad_top0, pad_right, pad_bottom0, pad_left,
      is_same_pad0, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, activation,
      prelu_weights_id, out_id);
  ASSERT_EQ(4, tfjs::backend::xnn_operator_count);

  // One more xnn operator should be created for the next call to conv2d with
  // bias, same input, same weights.
  tfjs::wasm::FusedDepthwiseConv2D(
      x0_id, batch_size, input_height, input_width, weights1_id, filter_height,
      filter_width, bias0_id, pad_top0, pad_right, pad_bottom0, pad_left,
      is_same_pad0, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, activation,
      prelu_weights_id, out_id);
  ASSERT_EQ(5, tfjs::backend::xnn_operator_count);

  // One more xnn operator should be created for the next call to conv2d with
  // new bias, same input, same weights.
  tfjs::wasm::FusedDepthwiseConv2D(
      x0_id, batch_size, input_height, input_width, weights1_id, filter_height,
      filter_width, bias1_id, pad_top0, pad_right, pad_bottom0, pad_left,
      is_same_pad0, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, activation,
      prelu_weights_id, out_id);
  ASSERT_EQ(6, tfjs::backend::xnn_operator_count);

  // One more xnn operator should be created for the next call to conv2d with
  // 'SAME' padding.
  const bool is_same_pad1 = true;
  tfjs::wasm::FusedDepthwiseConv2D(
      x0_id, batch_size, input_height, input_width, weights1_id, filter_height,
      filter_width, bias1_id, pad_top0, pad_right, pad_bottom0, pad_left,
      is_same_pad1, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, activation,
      prelu_weights_id, out_id);
  ASSERT_EQ(7, tfjs::backend::xnn_operator_count);

  // No new XNN operators should be created for the next call to conv2d with
  // 'SAME' padding and different input and raw padding values.
  tfjs::wasm::FusedDepthwiseConv2D(
      x1_id, batch_size, input_height, input_width, weights1_id, filter_height,
      filter_width, bias1_id, pad_top1, pad_right, pad_bottom1, pad_left,
      is_same_pad1, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, activation,
      prelu_weights_id, out_id);
  ASSERT_EQ(7, tfjs::backend::xnn_operator_count);

  // One new XNN operator should be created for the next call to conv2d with a
  // different activation.
  const int activation2 = tfjs::wasm::FusableActivation::RELU6;
  tfjs::wasm::FusedDepthwiseConv2D(
      x1_id, batch_size, input_height, input_width, weights1_id, filter_height,
      filter_width, bias1_id, pad_top1, pad_right, pad_bottom1, pad_left,
      is_same_pad1, dilation_height, dilation_width, stride_height,
      stride_width, input_channels, output_channels, activation2,
      prelu_weights_id, out_id);
  ASSERT_EQ(8, tfjs::backend::xnn_operator_count);

  // Disposing the first weights should remove 2 operators.
  tfjs::wasm::dispose_data(weights0_id);
  ASSERT_EQ(6, tfjs::backend::xnn_operator_count);

  // Disposing the second bias should remove 2 operators it's associated with.
  tfjs::wasm::dispose_data(bias1_id);
  ASSERT_EQ(3, tfjs::backend::xnn_operator_count);

  // Disposing the first bias should remove one operator.
  tfjs::wasm::dispose_data(bias0_id);
  ASSERT_EQ(2, tfjs::backend::xnn_operator_count);

  // Disposing the second weights should remove one operator.
  tfjs::wasm::dispose_data(weights1_id);
  ASSERT_EQ(1, tfjs::backend::xnn_operator_count);

  // Disposing prelu weights should remove the last operator.
  tfjs::wasm::dispose_data(prelu_weights_id);
  ASSERT_EQ(0, tfjs::backend::xnn_operator_count);

  tfjs::wasm::dispose();
}

File: tfjs-core/src/ops/fused/fused_depthwise_conv2d_test.ts
/**
 * @license
 * Copyright 2020 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */

import * as tf from '../../index';
import {ALL_ENVS, describeWithFlags} from '../../jasmine_util';
import {expectArraysClose} from '../../test_util';

describeWithFlags('fused depthwiseConv2D', ALL_ENVS, () => {
  it('basic', async () => {
    const fSize = 2;
    const pad = 'valid';
    const strides = 1;
    const chMul = 1;
    const inDepth = 1;

    const x = tf.tensor4d(
        [
          0.230664, 0.987388, 0.0685208, 0.419224, 0.887861, 0.731641,
          0.0741907, 0.409265, 0.351377
        ],
        [1, 3, 3, inDepth]);
    const w = tf.tensor4d(
        [-0.303873, -0.229223, 0.144333, 0.803373],
        [fSize, fSize, inDepth, chMul],
    );

    const result = tf.fused.depthwiseConv2d({x, filter: w, strides, pad});
    expect(result.shape).toEqual([1, 2, 2, 1]);
    const expected = [0.47737, 0.40018, 0.00859, -0.09615];
    expectArraysClose(await result.data(), expected);
  });

  it('basic with relu', async () => {
    const fSize = 2;
    const pad = 'valid';
    const strides = 1;
    const chMul = 1;
    const inDepth = 1;

    const x = tf.tensor4d(
        [
          0.230664, 0.987388, 0.0685208, 0.419224, 0.887861, 0.731641,
          0.0741907, 0.409265, 0.351377
        ],
        [1, 3, 3, inDepth]);
    const w = tf.tensor4d(
        [-0.303873, -0.229223, 0.144333, 0.803373],
        [fSize, fSize, inDepth, chMul],
    );

    const result = tf.fused.depthwiseConv2d(
        {x, filter: w, strides, pad, activation: 'relu'});
    expect(result.shape).toEqual([1, 2, 2, 1]);
    const expected = [0.47737, 0.40018, 0.00859, 0];
    expectArraysClose(await result.data(), expected);
  });

  it('basic with channel-wise broadcasted bias and relu', async () => {
    const strides = 1;
    const pad = 'same';
    const x = tf.tensor4d(
        [
          0, 1, 2, 3, 4, 5, 6, 7, 8, 0, 1, 2, 3, 4, 5, 6, 7, 8,
          0, 1, 2, 3, 4, 5, 6, 7, 8, 0, 1, 2, 3, 4, 5, 6, 7, 8
        ],
        [1, 3, 3, 4]);
    const w = tf.tensor4d(
        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], [2, 2, 4, 1]);
    const bias = tf.tensor1d([0, 1, 2, 3]);

    const result = tf.fused.depthwiseConv2d({x, filter: w, strides, pad, bias});
    expect(result.shape).toEqual([1, 3, 3, 4]);
    const expected = [
      124, 167, 92,  142, 112, 117, 76,  124, 16, 28, 44, 64,
      88,  134, 134, 88,  76,  120, 154, 205, 40, 58, 80, 106,
      4,   18,  36,  31,  20,  33,  50,  71,  0,  7,  16, 27
    ];
    expectArraysClose(await result.data(), expected);
  });

  it('basic with broadcasted bias and relu', async () => {
    const fSize = 2;
    const pad = 'valid';
    const strides = 1;
    const chMul = 1;
    const inDepth = 1;

    const x = tf.tensor4d(
        [
          0.230664, 0.987388, 0.0685208, 0.419224, 0.887861, 0.731641,
          0.0741907, 0.409265, 0.351377
        ],
        [1, 3, 3, inDepth]);
    const w = tf.tensor4d(
        [-0.303873, -0.229223, 0.144333, 0.803373],
        [fSize, fSize, inDepth, chMul],
    );

    const result = tf.fused.depthwiseConv2d(
        {x, filter: w, strides, pad, bias: tf.scalar(1), activation: 'relu'});
    expect(result.shape).toEqual([1, 2, 2, 1]);
    const expected = [1.47737, 1.40018, 1.00859, 0.90385];
    expectArraysClose(await result.data(), expected);
  });

  it('prelu', async () => {
    const fSize = 3;
    const pad = 'valid';
    const strides = 1;
    const chMul = 1;
    const inDepth = 1;

    const x = tf.tensor4d(
        [
          0.149194, 0.089009, 0.654891, 0.083324, 0.537043, 0.644331, 0.563037,
          0.211859, 0.633501, 0.186427, 0.777034, 0.50001,  0.607341, 0.95303,
          0.696479, 0.050387, 0.62045,  0.728049, 0.028043, 0.437009, 0.712881,
          0.741935, 0.974474, 0.621102, 0.171411
        ],
        [1, 5, 5, inDepth]);
    const alpha = tf.tensor4d(
        [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9], [1, 3, 3, 1]);
    const w = tf.tensor4d(
        [
          -0.125386, -0.975199, -0.640437, -0.281895, -0.990968, -0.347208,
          -0.889702, -0.180695, -0.691992
        ],
        [fSize, fSize, inDepth, chMul],
    );

    const result = tf.fused.depthwiseConv2d({
      x,
      filter: w,
      strides,
      pad,
      activation: 'prelu',
      preluActivationWeights: alpha
    });
    expect(result.shape).toEqual([1, 3, 3, 1]);
    const expected = [
      -0.25400, -0.50118, -0.73622, -0.94068, -1.2298, -1.84585, -2.3089,
      -2.7499, -2.64077
    ];
    expectArraysClose(await result.data(), expected);
  });

  it('leakyrelu', async () => {
    const fSize = 3;
    const pad = 'valid';
    const strides = 1;
    const chMul = 1;
    const inDepth = 1;

    const x = tf.tensor4d(
        [
          0.149194, 0.089009, 0.654891, 0.083324, 0.537043, 0.644331, 0.563037,
          0.211859, 0.633501, 0.186427, 0.777034, 0.50001,  0.607341, 0.95303,
          0.696479, 0.050387, 0.62045,  0.728049, 0.028043, 0.437009, 0.712881,
          0.741935, 0.974474, 0.621102, 0.171411
        ],
        [1, 5, 5, inDepth]);
    const alpha = 0.3;
    const w = tf.tensor4d(
        [
          -0.125386, -0.975199, -0.640437, -0.281895, -0.990968, -0.347208,
          -0.889702, -0.180695, -0.691992
        ],
        [fSize, fSize, inDepth, chMul],
    );

    const result = tf.fused.depthwiseConv2d({
      x,
      filter: w,
      strides,
      pad,
      activation: 'leakyrelu',
      leakyreluAlpha: alpha
    });

    expect(result.shape).toEqual([1, 3, 3, 1]);
    const expected = [
      -0.7620067596435547, -0.7517655491828918, -0.7362186312675476,
      -0.7055101990699768, -0.7378802299499512, -0.9229262471199036,
      -0.9895440340042114, -1.031226396560669, -0.8802568912506104
    ];
    expectArraysClose(await result.data(), expected);
  });

  it('sigmoid', async () => {
    const fSize = 3;
    const pad = 'valid';
    const strides = 1;
    const chMul = 1;
    const inDepth = 1;

    const x = tf.tensor4d(
        [
          0.149194, 0.089009, 0.654891, 0.083324, 0.537043, 0.644331, 0.563037,
          0.211859, 0.633501, 0.186427, 0.777034, 0.50001,  0.607341, 0.95303,
          0.696479, 0.050387, 0.62045,  0.728049, 0.028043, 0.437009, 0.712881,
          0.741935, 0.974474, 0.621102, 0.171411
        ],
        [1, 5, 5, inDepth]);
    const w = tf.tensor4d(
        [
          -0.125386, -0.975199, -0.640437, -0.281895, -0.990968, -0.347208,
          -0.889702, -0.180695, -0.691992
        ],
        [fSize, fSize, inDepth, chMul],
    );

    const result = tf.fused.depthwiseConv2d(
        {x, filter: w, strides, pad, activation: 'sigmoid'});

    expect(result.shape).toEqual([1, 3, 3, 1]);
    const expected = [
      0.07309964, 0.07544667, 0.07914197, 0.08693069, 0.07873929, 0.04409045,
      0.03562334, 0.0311462, 0.05048907
    ];
    expectArraysClose(await result.data(), expected);
  });

  it('gradient x=[2,3,3,1] f=[2,2,1,1] s=1 p=0', async () => {
    const inputDepth = 1;
    const outputDepth = 1;
    const inputShape: [number, number, number, number] = [2, 3, 3, inputDepth];
    const filterSize = 2;
    const strides = 1;
    const pad = 0;

    const filterShape: [number, number, number, number] =
        [filterSize, filterSize, inputDepth, outputDepth];
    const filter = tf.tensor4d([-1, 1, -2, 0.5], filterShape);

    const x = tf.tensor4d(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 6, 7, 8, 9], inputShape);
    const dy = tf.tensor4d([3, 1, 2, 0, 3, 1, 2, 0], [2, 2, 2, 1]);

    const grads = tf.grads(
        (x: tf.Tensor4D, filter: tf.Tensor4D) =>
            tf.fused.depthwiseConv2d({x, filter, strides, pad}));
    const [dx, dfilter] = grads([x, filter], dy);

    expect(dx.shape).toEqual(x.shape);
    expectArraysClose(
        await dx.data(),
        [-3, 2, 1, -8, 1.5, 0.5, -4, 1, 0, -3, 2, 1, -8, 1.5, 0.5, -4, 1, 0]);

    expect(dfilter.shape).toEqual(filterShape);
    expectArraysClose(await dfilter.data(), [26, 38, 62, 74]);
  });

  it('gradient x=[2,3,3,1] f=[2,2,1,1] s=1 p=0 with bias', async () => {
    const inputDepth = 1;
    const outputDepth = 1;
    const inputShape: [number, number, number, number] = [2, 3, 3, inputDepth];
    const filterSize = 2;
    const strides = 1;
    const pad = 0;

    const filterShape: [number, number, number, number] =
        [filterSize, filterSize, inputDepth, outputDepth];
    const filter = tf.tensor4d([-1, 1, -2, 0.5], filterShape);
    const bias = tf.ones([2, 2, 2, 1]);

    const x = tf.tensor4d(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 6, 7, 8, 9], inputShape);
    const dy = tf.tensor4d([3, 1, 2, 0, 3, 1, 2, 0], [2, 2, 2, 1]);

    const fusedGrads = tf.grads(
        (x: tf.Tensor4D, w: tf.Tensor4D, b) => tf.fused.depthwiseConv2d({
          x,
          filter: w,
          strides,
          pad,
          dataFormat: 'NHWC',
          dilations: [1, 1],
          bias: b
        }));
    const [dxFused, dfilterFused, dbiasFused] =
        fusedGrads([x, filter, bias], dy);

    const grads = tf.grads((x: tf.Tensor4D, filter: tf.Tensor4D, bias) => {
      const conv = tf.depthwiseConv2d(x, filter, strides, pad);
      const sum = tf.add(conv, bias);
      return sum;
    });
    const [dx, dfilter, dbias] = grads([x, filter, bias], dy);

    expectArraysClose(await dxFused.array(), await dx.array());
    expectArraysClose(await dfilterFused.array(), await dfilter.array());
    expectArraysClose(await dbiasFused.array(), await dbias.array());
  });

  it('gradient x=[2,3,3,1] f=[2,2,1,1] s=1 p=0 with bias and activation',
     async () => {
       const inputDepth = 1;
       const outputDepth = 1;
       const inputShape: [number, number, number, number] =
           [2, 3, 3, inputDepth];
       const filterSize = 2;
       const strides = 1;
       const pad = 0;

       const filterShape: [number, number, number, number] =
           [filterSize, filterSize, inputDepth, outputDepth];
       const filter = tf.tensor4d([-1, 1, -2, 0.5], filterShape);
       const bias = tf.ones([2, 2, 2, 1]);

       const x = tf.tensor4d(
           [1, 2, 3, 4, 5, 6, 7, 8, 9, 1, 2, 3, 4, 5, 6, 7, 8, 9], inputShape);
       const dy = tf.tensor4d([3, 1, 2, 0, 3, 1, 2, 0], [2, 2, 2, 1]);

       const fusedGrads = tf.grads(
           (x: tf.Tensor4D, w: tf.Tensor4D, b) => tf.fused.depthwiseConv2d({
             x,
             filter: w,
             strides,
             pad,
             dataFormat: 'NHWC',
             dilations: [1, 1],
             bias: b,
             activation: 'relu'
           }));
       const [dxFused, dfilterFused, dbiasFused] =
           fusedGrads([x, filter, bias], dy);

       const grads = tf.grads((x: tf.Tensor4D, filter: tf.Tensor4D, bias) => {
         const conv = tf.depthwiseConv2d(x, filter, strides, pad);
         const sum = tf.add(conv, bias);
         return tf.relu(sum);
       });
       const [dx, dfilter, dbias] = grads([x, filter, bias], dy);

       expectArraysClose(await dxFused.array(), await dx.array());
       expectArraysClose(await dfilterFused.array(), await dfilter.array());
       expectArraysClose(await dbiasFused.array(), await dbias.array());
     });
});
